Reset counter mpe

Zethian opened the pull request:
Adds counter that increments on every restart of the system. Can be reset with the syscommand 0x55 (MSP_OP_CUBES_RST_RST)

Commit:
Add reset counter functionality

Adds functions that increments a NVM counter whenever the system is
booted.

Has a read function which adds the uint32_t value to the end of
housekeeping data.

Has a reset function which sets the counter to 0 and is called by the
OBC as a syscommand with OPcode 0x55.

NB: These changes are untested.

Commit:
Fix bugs in reset counter

Fixed bugs and tested on system so it works.

tstana commented:
I should be able to check and test within tomorrow afternoon. I can do the merge (or give further comments) at that point.

tstana commented:
Thanks for this work, it looks good!

Please also implement the following before we can merge this PR:

- make MSP_OP_CUBES_RST_RST a SEND command, renaming it to MSP_OP_CUBES_SEND_RST
- add a single byte as parameter to the new MSP_OP_CUBES_SEND_RST command; the byte shall contain the following information:

bit 0: when set, call the nvm_reset_counter_reset() function
bit 1: when set, call the citiroc_hcr_reset() function
bit 2: when set, call the citiroc_histo_reset() function
bit 3: when set, call a commented out "citiroc_psc_reset()" function
bit 4: when set, call a commented out "citiroc_sr_reset()" function
bit 5: when set, call a commented out "citiroc_pa_reset()" function
bit 6: when set, call a commented out "citiroc_or32_reset()" function
bit 7: no action

Commit:
Move reset to send command. Use for more functions

Moves the usage of the reset command to a send command to accept one
byte. This byte will have different bits set depending on which
resets are needed.
A if tree checks what resets should be called. One bit is reserved
for later usage.

bit resets:
0: reset counter
1: hcr counter
2: histogram
3: psc
4: sr
5: pa
6: or32
7: Reserved

Commit:
Add more Citiroc reset functions

tstana commented:
I cannot seem to read reliable values on the reset counter from the CitirocUI. The issue is not with the CitirocUI, it is in the cubes-fw. I ran debug so far as to 

1. Check that the value received by REQ_HK is received on the CitirocUI;
2. Sending a SEND_CUBES_RESET command (MSP through CitirocUI) with the appropriate bit set runs the nvm_reset_counters_reset() function as it should.

I did not however check the contents of the memory itself. What I could see -- from the CitirocUI end -- is that:

1. Each REQ_HK yields the same value read out consistently (reset counters has no effect)
2. Power-cycling does not increment the counters

On a quick look through your code, the issue could be because of the mismatch between uint32_t and uint8_t at **line 138** in mem_mgmt.c:

```
	mem_nvm_write(NVM_RESET, (uint8_t *)&counter);
```

Steps to reproduce:

1. Start CitirocUI (commit id `60fe0cc2`)
2. Connect to the CUBES board (I attempted with Rev. B board, but Rev. A should behave the same)
3. Pull out the CUBES Monitor and run "Read Telemetry"
4. Check the value of the reset counter
5. Power-cycle the CUBES board
6. "Read Telemetry" again
7. Observe that the reset counter has not incremented
8. Tick the "Reset Counters" check-box in CitirocUI
9. Click the "Send Resets" button
10. "Read Telemetry"
11. Observe that the reset counter has not reset (if it even was read as different from '0') to begin with.

Commit:
Implement and uncomment reset functions in main.c

Without uncommenting these functions, an interesting feature was added
by the compiler, whereby the contents of the if branch following an if
branch with commented contents were not executed. For example, the code
under the second if branch was not executed:

  if(resetvalue & 0b00010000)
    citiroc_sr_reset();
  if(resetvalue & 0b00100000)
    citiroc_pa_reset();

Commit:
Add HVPS reset function

Adds a function for resetting the output for the HVPS

listed at bit 7 on the reset functionality.

Commit:
Add Citiroc read reg reset

My version of the code contained uncommitted changes, which were also
using bit 7 of the MSP_OP_SEND_CUBES_RST command parameter. Offsetting
the HVPS reset to a separate byte in the command parameter, and adding
code in citiroc.c, .h for the read register reset.
